### PR TITLE
Casmcms 7611

### PR DESCRIPTION
**Pull request by jsl-hpe**
## Summary and Scope

Included are build changes required to comply with updated base charts that include accurate image references, as well as changes to the way BOS is packaged as a helm chart. This includes changes that allow for local testing of the BOS charts before pushing code to github for build tests.

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* CASMCMS-7611
* Changes will need to be promoted to master as well as release/1.2

## Testing

- Mug & Rocket (where available); HELM update tests using a manifest to consume charts from a local directory <path>.

### Tested on:

  * mug
  * rocket

### Test description:

loftsman update commands were used with a manifest file to test changes.

## Risks and Mitigations

The base charts and underlying build repositories are always changing day to day. What works now may not work later, but by no fault of this particular commit.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Target branch correct
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,14 +55,31 @@ TEST_SPEC_FILE ?= ${TEST_SPEC_NAME}.spec
...
+		cp bos-reporter.spec.in bos-reporter.spec
+		cp bos-crayctldeploy-test.spec.in bos-crayctldeploy-test.spec
+		cp src/setup.py.in src/setup.py
+		cp api/openapi.yaml.in api/openapi.yaml
+
+cms_meta_tools:
+		rm -rf cms-meta-tools
+		git clone --depth 1 --no-single-branch git@github.com:Cray-HPE/cms-meta-tools.git ./cms_meta_tools
+
+runbuildprep:   clone_input_files
```
Comment (jsollom-hpe) [inline Makefile:76]: Nit: clone_input_files should be on its own line to be consistent with the other formatting.

File: kubernetes/cray-bos/Chart.yaml.in [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    - name: cray-bos
+      image: artifactory.algol60.net/csm-docker/stable/cray-bos:0.0.0-bos
+    - name: redis
+      image: docker.io/library/redis:5.0-alpine3.14
+  artifacthub.io/license: MIT
+apiVersion: v2
+appVersion: 0.0.0
+dependencies:
+- name: cray-service
+  version: ^7.0.0
```
Comment (jsollom-hpe) [inline kubernetes/cray-bos/Chart.yaml.in:14]: I applaud the use of '^'. I think it's important to be aggressive on accepting these changes.

File: kubernetes/cray-bos/Chart.yaml.in [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+keywords:
+  - boa
+  - bos
+  - cray-bos
+maintainers:
+  - name: Jason
+    email: jason.sollom@hpe.com
+  - name: Joel
+    email: joel.landsteiner@hpe.com
+  - name: Ryan
```
Comment (jsollom-hpe) [inline kubernetes/cray-bos/Chart.yaml.in:27]: Rob supplanted by Ryan. Conservation of Rs, I guess.

File: kubernetes/cray-bos/values.yaml.in [hunk 4]
```diff
@@ -8,9 +8,12 @@
 #   pullPolicy: "" (default = "IfNotPresent")
 
 boa_image:
-  repository: cray/cray-boa
-  tag: IMAGE_CRAY_BOA_TAG
-
+  repository: artifactory.algol60.net/csm-docker/stable/cray-boa
```
Comment (jsollom-hpe) [inline kubernetes/cray-bos/values.yaml.in:11]: Do we want boa_image to have an 'image' attribute to be symmetric with 'database'?

File: kubernetes/cray-bos/values.yaml.in [hunk 5]
```diff
@@ -34,10 +37,11 @@ cray-service:
     cray-bos:
       name: cray-bos
       image:
-        repository: cray/cray-bos
+        repository: artifactory.algol60.net/csm-docker/stable/cray-bos
+        # tag defaults to chart appVersion
       ports:
       - name: http
-        containerPort: 8000
+        containerPort: 9000
```
Comment (jsollom-hpe) [inline kubernetes/cray-bos/values.yaml.in:44]: Why are we changing the containerPort?

File: update_versions.conf [hunk 6]
```diff
@@ -30,17 +30,24 @@ targetfile: src/setup.py
 # The following file does not exist in the repo as a static file
 # It is generated at build time
 sourcefile: .chart_version
-tag: @CHART_VERSION@
+tag: 0.0.0-chart
```
Comment (jsollom-hpe) [inline update_versions.conf:33]: Crud, so now we have to remember to increment all of these versions when we bump a version?

File: .chart_version [hunk 7]
```diff
@@ -0,0 +1 @@
+0.0.0
```
[the remark this answers is not in the document]
Comment (jsl-hpe) [reply .chart_version:1]: We're really not, but the way buildprep.sh and associated makefiles behave, a valid version field needs to exist. As a result, this file allows you to make the build locally outside of the build pipeline. They are functionally overwritten with the correct version in our build pipeline.

File: .github/workflows/charts-lint-test-scan.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+name: Lint, test, and scan Helm charts
+on:
+  pull_request:
+    branches:
+      - master
+      - release/**
+  schedule:
+    - cron: "0 0 * * *"
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (jsl-hpe) [reply .github/workflows/charts-lint-test-scan.yml:9]: yes.

File: kubernetes/cray-bos/Chart.yaml.in (hunk 3 above)
Comment (jsl-hpe) [reply kubernetes/cray-bos/Chart.yaml.in:27]: There must be balance.

File: kubernetes/cray-bos/values.yaml.in (hunk 4 above)
Comment (jsl-hpe) [reply kubernetes/cray-bos/values.yaml.in:11]: Given that BOA is going away, I don't particularly care much.

File: update_versions.conf (hunk 6 above)
Comment (jsl-hpe) [reply update_versions.conf:33]: No. The build pipeline and buildprep.sh inject the correct values into these files (or the copies of the files). We still have to increment and release with a tag.

File: Makefile [hunk 9]
```diff
@@ -54,14 +55,31 @@ TEST_SPEC_FILE ?= ${TEST_SPEC_NAME}.spec
 TEST_SOURCE_NAME ?= ${TEST_SPEC_NAME}-${RPM_VERSION}
 TEST_SOURCE_PATH := ${TEST_BUILD_DIR}/SOURCES/${TEST_SOURCE_NAME}.tar.bz2
 
-all : runbuildprep lint prepare image chart rptr_rpm test_rpm
+all : runbuildprep lint image chart rptr_rpm test_rpm
```
[the remark this answers is not in the document]
Comment (jsl-hpe) [reply Makefile:58]: because prepare is no longer a build step.

File: kubernetes/cray-bos/Chart.yaml.in [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  artifacthub.io/images: |
+    - name: cray-boa
+      image: artifactory.algol60.net/csm-docker/stable/cray-boa:0.0.0-boa
+    - name: cray-bos
+      image: artifactory.algol60.net/csm-docker/stable/cray-bos:0.0.0-bos
+    - name: redis
+      image: docker.io/library/redis:5.0-alpine3.14
+  artifacthub.io/license: MIT
+apiVersion: v2
+appVersion: 0.0.0
```
Comment (dlaine-hpe) [inline kubernetes/cray-bos/Chart.yaml.in:11]: Should this now be '0.0.0-bos'?  I don't see a '0.0.0' tag in the update_versions.conf file...

**Review by dlaine-hpe — CHANGES_REQUESTED** [review]

File: kubernetes/cray-bos/Chart.yaml.in (hunk 10 above)
Comment (jsl-hpe) [reply kubernetes/cray-bos/Chart.yaml.in:11]: Nope; this is replaced by cms-meta-tools.

File: kubernetes/cray-bos/values.yaml.in (hunk 5 above)
Comment (jsl-hpe) [reply kubernetes/cray-bos/values.yaml.in:44]: Master branch is apparently lacking non-root mods. We should open a ticket to track to see if there is anything further missing, or submit a PR 1.2 --- merge --> master to capture everything else that may be missing.

File: Makefile (hunk 1 above)
Comment (jsl-hpe) [reply Makefile:76]: The formatting is consistent, but the diff view doesn't make that particularly apparent:

[snip]
test_rpm: test_rpm_package_source test_rpm_build_source test_rpm_build

clone_input_files:
		cp ${CHART_PATH}/${NAME}/Chart.yaml.in ${CHART_PATH}/${NAME}/Chart.yaml
		cp ${CHART_PATH}/${NAME}/values.yaml.in ${CHART_PATH}/${NAME}/values.yaml
		cp bos-reporter.spec.in bos-reporter.spec
		cp bos-crayctldeploy-test.spec.in bos-crayctldeploy-test.spec
		cp src/setup.py.in src/setup.py
		cp api/openapi.yaml.in api/openapi.yaml

cms_meta_tools:
		rm -rf cms-meta-tools
[snip

File: kubernetes/cray-bos/Chart.yaml.in (hunk 3 above)
Comment (rbak-hpe) [reply kubernetes/cray-bos/Chart.yaml.in:27]: According to Zach the maintainer names must be the our hpe github usernames

Comment (jsl-hpe) [reply kubernetes/cray-bos/Chart.yaml.in:27]: This should be documented somehwere we can reference it.

File: kubernetes/cray-bos/values.yaml.in [hunk 11]
```diff
@@ -8,9 +8,12 @@
...
 
 boa_image:
-  repository: cray/cray-boa
-  tag: IMAGE_CRAY_BOA_TAG
-
+  repository: artifactory.algol60.net/csm-docker/stable/cray-boa
+  tag: 0.0.0-boa 
+database:
+  image:
+    repository: docker.io/library/redis
```
Comment (rkleinman-hpe) [inline kubernetes/cray-bos/values.yaml.in:15]: This should be coming from artifactory.algol60.net

**Review by rkleinman-hpe — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 12]
```diff
@@ -21,8 +21,12 @@
 #
 # (MIT License)
 
+# Upstream Build Args
+ARG OPENAPI_IMAGE=openapitools/openapi-generator-cli:v4.1.2
```
Comment (ecozzi-hpe) [inline Dockerfile:25]: Shouldn't this be pulled from algol60?

File: kubernetes/cray-bos/values.yaml.in (hunk 5 above)
Comment (jsollom-hpe) [reply kubernetes/cray-bos/values.yaml.in:44]: I created CASMCMS-7696 to do the non-root audit. 
Why is port 8000 considered a non-root change? Typically only ports 1024 and under are restricted to root for the Linux kernel.

Comment (jsl-hpe) [reply kubernetes/cray-bos/values.yaml.in:44]: uwsgi cannot bind to port 8000 when launched as non-root.

File: kubernetes/cray-bos/values.yaml.in (hunk 11 above)
Comment (jsl-hpe) [reply kubernetes/cray-bos/values.yaml.in]: It is now.

File: Dockerfile (hunk 12 above)
Comment (jsl-hpe) [reply Dockerfile:25]: The version of openapi generator in algol60 is not compatible with BOS spec.

Comment (ecozzi-hpe) [reply Dockerfile:25]: Then we should get a compatible version on algol60.

**Review by rbak-hpe — APPROVED** [review]

File: Dockerfile (hunk 12 above)
Comment (jsl-hpe) [reply Dockerfile:25]: That's fine, but not the point of this mod. https://connect.us.cray.com/jira/browse/CASMCMS-7697

**Review by jsollom-hpe — APPROVED** [review]

**Review by dlaine-hpe — APPROVED** [review]

**Review by rkleinman-hpe — APPROVED** [review]

**Pull request merged**